### PR TITLE
enh: move codec tests

### LINKED ISSUE #11536: Move OP codecs test
### Describe the feature

Move OP codecs test into `reth-optimism-storage`

https://github.com/paradigmxyz/reth/blob/6a1ec0d1a20436a79aee612c68749abc3411ec6b/crates/storage/codecs/src/alloy/transaction/mod.rs#L34-L38

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by edisontim**
closes #11536

### CODE REVIEW

File: crates/storage/codecs/src/alloy/transaction/mod.rs [hunk 1]
```diff
@@ -14,26 +14,4 @@ mod tests {
     //
     // this check is to ensure we do not inadvertently add too many fields to a struct which would
     // expand the flags field and break backwards compatibility
-
-    #[cfg(feature = "optimism")]
-    use crate::alloy::transaction::optimism::TxDeposit;
-    use crate::alloy::transaction::{
-        eip1559::TxEip1559, eip2930::TxEip2930, eip4844::TxEip4844, eip7702::TxEip7702,
-        legacy::TxLegacy,
-    };
```
Comment (emhane) [inline crates/storage/codecs/src/alloy/transaction/mod.rs:23]: these imports should stay here, note that they don't have an optimism feature gate

File: crates/optimism/storage/src/lib.rs [hunk 2]
```diff
@@ -11,6 +11,10 @@
 
 #[cfg(test)]
 mod tests {
+    use crate::alloy::transaction::{
+        eip1559::TxEip1559, eip2930::TxEip2930, eip4844::TxEip4844, eip7702::TxEip7702,
+        legacy::TxLegacy,
+    };
```
Comment (emhane) [inline crates/optimism/storage/src/lib.rs:17]:
```suggestion
```
we don't need this here

File: crates/optimism/storage/src/lib.rs [hunk 3]
```diff
@@ -75,5 +79,20 @@ mod tests {
...
         validate_bitflag_backwards_compat!(Withdrawals, UnusedBits::Zero);
         validate_bitflag_backwards_compat!(Requests, UnusedBits::Zero);
+
+        assert_eq!(TxEip4844::bitflag_encoded_bytes(), 5);
+        assert_eq!(TxLegacy::bitflag_encoded_bytes(), 3);
+        assert_eq!(TxEip1559::bitflag_encoded_bytes(), 4);
+        assert_eq!(TxEip2930::bitflag_encoded_bytes(), 3);
+        assert_eq!(TxEip7702::bitflag_encoded_bytes(), 4);
+    }
+
```
Comment (emhane) [inline crates/optimism/storage/src/lib.rs:89]:
```suggestion
```
we don't need this here

**Review by emhane — CHANGES_REQUESTED** [review]

**Comment by edisontim** [comment]
Hello @emhane, I believe this can't be done at the moment because the alloy module of codecs isn't declared as public, neither is the transaction module inside the alloy, and to move the codec tests into the optimism crate I need to access `TxDeposit`.
I imagine it's planned to move `TxDeposit` into the optimism crate but in the meantime I believe this PR can be closed

**Comment by emhane** [comment]
oh yeah this one is tricky, will push a fix

**Review by emhane — APPROVED** [review]

**Comment by edisontim** [comment]
Awesome, thanks for that. Nothing else needed on my end on this PR right?

File: crates/primitives/src/transaction/mod.rs [hunk 4]
```diff
@@ -715,6 +715,7 @@ impl reth_codecs::Compact for Transaction {
             }
             #[cfg(feature = "optimism")]
             Self::Deposit(tx) => {
+                let tx: reth_optimism_codecs::TxDepositEncode<'_> = tx.into();
```
Comment (mattsse) [inline crates/primitives/src/transaction/mod.rs:718]: why is this necessary, seems odd?

**Review by mattsse — COMMENTED** [review]
wdyt @joshieDo

File: crates/optimism/codecs/src/tx_deposit.rs [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+///
+/// This is a helper type for encoding. It allows for using derive on the type instead of manually
+/// managing `bitfield`.
+///
+/// By deriving `Compact` here, any future changes or enhancements to the `Compact` derive
+/// will automatically apply to this type.
+///
+/// Notice: Asides references, should match struct [`op_alloy_consensus::TxDeposit`]
+#[derive(Debug, Compact)]
+pub struct TxDepositEncode<'a> {
```
[the remark this answers is not in the document]
Comment (emhane) [reply crates/optimism/codecs/src/tx_deposit.rs:18]: because can't impl `Compact` trait on alloy type (foreign type) outside of crate either is defined

File: crates/primitives/src/transaction/mod.rs (hunk 4 above)
Comment (emhane) [reply crates/primitives/src/transaction/mod.rs:718]: intermediary type needed cause can't impl `Compact` on op-alloy type in crate where `Compact` is not defined

**Comment by edisontim** [comment]
Hey :) Can this be merged?

File: crates/storage/codecs/src/alloy/transaction/optimism.rs
Comment (mattsse) [inline crates/storage/codecs/src/alloy/transaction/optimism.rs:1]: lets keep this, this is much nicer than having to use an intermediary type for no reason

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/primitives/src/transaction/mod.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (edisontim) [reply crates/primitives/src/transaction/mod.rs:718]: Hey @mattsse, should I close this PR then? Maybe I'm missing something but I'm not sure I can move the codecs tests without making the `TxDeposit` struct public outside of its crate

[the remark this answers is not in the document]
Comment (emhane) [reply crates/primitives/src/transaction/mod.rs:718]: yes exactly, then the optimism feature can't be removed if this type stays in the compact crate, which creates a bad patter for further rollups

[the remark this answers is not in the document]
Comment (mattsse) [reply crates/primitives/src/transaction/mod.rs:718]: this optimism feature would be harmless because this is doesn't interfere with any other code.
and it's a lot cleaner than introducing an intermediary type for no reason

**Comment by mattsse** [comment]
closing this because is just makes it more complex

**Comment by emhane** [comment]
I don't see a way this complexity can be avoided @mattsse , ref https://github.com/paradigmxyz/reth/pull/11549/files#r1800367074

**Comment by emhane** [comment]
> I don't understand what this solves
> 
> what's the benefit here? why not keep the nativel compact impl?

because we don't want to have the optimism feature in that crate

**Comment by emhane** [comment]
> > I don't understand what this solves
> > what's the benefit here? why not keep the nativel compact impl?
> 
> because we don't want to have the optimism feature in that crate @mattsse
